File: php/src/CurrencyConverter.php
<?php

namespace kata;

use kata\external\ConversionRateApi;
use kata\model\Currency;
use kata\model\Money;

class CurrencyConverter
{
    private ConversionRateApi $conversionRateApi;

    public function __construct(ConversionRateApi $api)
    {
        $this->conversionRateApi = $api;
    }

    /**
     * @throws \Exception
     */
    function sum(Currency $target, array $moneys) : Money
    {
        $total = 0;
        $knownRates = [$target->toString() => 1];

        /* @var $m Money */
        foreach($moneys as $m){
            if (!array_key_exists($m->currency->toString(), $knownRates)) {
                $knownRates[$m->currency->toString()] = $this->conversionRateApi->getRate($m->currency->toIsoCurrency(), $target->toIsoCurrency());
            }

            $total = $knownRates[$m->currency->toString()] * $m->amount;
        }

        return new Money($total, $target);
    }

}

File: python/src/currency_converter.py
from src.currency import Currency
from src.external.conversion_rate_api import ConversionRateApi
from src.money import Money


class CurrencyConverter:
    def __init__(self, rate_api: ConversionRateApi):
        self.rate_api = rate_api

    def sum(self, target: Currency, *moneys: Money):
        amount = 0.0

        known_rates = {target: 1.0}

        for money in moneys:
            if money.currency not in known_rates:
                known_rates[money.currency] = self.rate_api.get_rate(money.currency.to_iso_code(), target.to_iso_code())
            amount += money.amount * known_rates[money.currency]

        return Money(amount, target)
